### PR TITLE
CV-192 CV-193 CV-312

### PR DESCRIPTION
**Pull request by ChughAkshay**

### CODE REVIEW

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Validators/WhenValidatingSelectProviderViewModel.cs [hunk 1]
```diff
@@ -7,11 +8,28 @@ namespace SFA.DAS.EmployerCommitmentsV2.Web.UnitTests.Validators
...
+            SelectProviderViewModel viewModel,
+            SelectProviderViewModelValidator validator)
+        {
+            viewModel.ProviderId = "abcdefghijklmnopqrstuvwxyz*!£$%^&*()_+¬`=-][#';/.,}{~@:?><";
+            var result = validator.Validate(viewModel);
+
+            Assert.False(result.IsValid);
+        }
 
         [Test, MoqAutoData]
```
Comment (scottywakefield) [inline src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Validators/WhenValidatingSelectProviderViewModel.cs]: this test can now be removed or renamed, "0" isn't a default value.

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs [hunk 2]
```diff
@@ -35,20 +53,41 @@ public IActionResult Index(IndexRequest request)
...
+
+                return RedirectToAction("ConfirmProvider", confirmProviderRequest);
+            }
+            catch (RestHttpClientException e)
+            {
+                if (e.StatusCode == HttpStatusCode.NotFound)
+                {
+                    ModelState.AddModelError(nameof(providerResponse.ProviderId), "Check UK Provider Reference Number");
+                    return View(request);
+                }
```
Comment (scottywakefield) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs]: else ... ?? other RestHttpClientExceptions won't get logged or anything, which would not be good. also worth having another catch for exception and redirect to the error page.

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Validators/SelectProviderViewModelValidator.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    public class SelectProviderViewModelValidator : AbstractValidator<SelectProviderViewModel>
+    {
+        public SelectProviderViewModelValidator()
+        {
+            RuleFor(x => x.ProviderId)
+                .Must(BeALongGreaterThanZero)
+                .WithMessage("Check UK Provider Reference Number");
+            RuleFor(x => x.EmployerAccountLegalEntityPublicHashedId)
+                .NotNull()
+                .NotEmpty();
```
Comment (scottywakefield) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Validators/SelectProviderViewModelValidator.cs]: do you need a message for this one?

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Validators/SelectProviderViewModelValidator.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace SFA.DAS.EmployerCommitmentsV2.Web.Validators
+{
+    public class SelectProviderViewModelValidator : AbstractValidator<SelectProviderViewModel>
+    {
+        public SelectProviderViewModelValidator()
+        {
+            RuleFor(x => x.ProviderId)
+                .Must(BeALongGreaterThanZero)
+                .WithMessage("Check UK Provider Reference Number");
```
Comment (scottywakefield) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Validators/SelectProviderViewModelValidator.cs]: you don't have any tests to verify this message.

File: src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Validators/WhenValidatingSelectProviderViewModel.cs (hunk 1 above)
Comment (ChughAkshay) [reply src/SFA.DAS.EmployerCommitmentsV2.Web.UnitTests/Validators/WhenValidatingSelectProviderViewModel.cs]: Removed test

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs (hunk 2 above)
Comment (ChughAkshay) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs]: Added catch for unexpected exceptions which log and return the error view

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Validators/SelectProviderViewModelValidator.cs (hunk 4 above)
Comment (ChughAkshay) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Validators/SelectProviderViewModelValidator.cs]: added unit test to verify error

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs [hunk 5]
```diff
@@ -73,21 +80,26 @@ public async Task<IActionResult> SelectProvider(SelectProviderViewModel request)
...
                 }
+                    _logger.LogError($"Failed '{nameof(CreateCohortController)}-{nameof(SelectProvider)}': {nameof(ex.StatusCode)}='{ex.StatusCode}', {nameof(ex.ReasonPhrase)}='{ex.ReasonPhrase}'");
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError($"Failed '{nameof(CreateCohortController)}-{nameof(SelectProvider)}': {nameof(ex.Message)}='{ex.Message}', {nameof(ex.StackTrace)}='{ex.StackTrace}'");
             }
 
-            return View(request);
+            return View("~/Views/Error/Error.cshtml");
```
Comment (scottywakefield) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs]: use RedirectToAction("Error", "Error") instead. in your test you can verify the ActionName and ControllerName.

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortWithDraftApprenticeshipController.cs [hunk 6]
```diff
@@ -82,20 +87,29 @@ public async Task<IActionResult> AddDraftApprenticeship(AddDraftApprenticeshipVi
...
         }
 
         private Task<IReadOnlyList<ITrainingProgramme>> GetCourses()
         {
             return _trainingProgrammeApiClient.GetAllTrainingProgrammes();
         }
+
+        private async Task<string> GetProviderName(long providerId)
+        {
+            var blah = await _commitmentsApiClient.GetProvider(providerId);
```
Comment (scottywakefield) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortWithDraftApprenticeshipController.cs]: noooop! rename plz.

Comment (ChughAkshay) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortWithDraftApprenticeshipController.cs]: This change appears to have come as a result of a merge fail when pulling master into CV-192 resulting in commit 068e4c3b74313e348a419a7d0add1cce12f1515e being skipped. I've hard reset the branch, pulled from master and force pushed, which solved the issue.

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Validators/SelectProviderViewModelValidator.cs (hunk 3 above)
Comment (ChughAkshay) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Validators/SelectProviderViewModelValidator.cs]: This value is set from the query string and isn't something the user explicitly provides. The search box only accepts a provider id and only shows validation for the provider id based on the screen shot.

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs [hunk 7]
```diff
@@ -35,20 +57,55 @@ public IActionResult Index(IndexRequest request)
...
         {
-            //todo:hit api
+            if (!ModelState.IsValid)
+            {
+                return RedirectToAction("Error", "Error", new { StatusCode = 400 });
+            }
+
+            try
+            {
+                var validationResult = _selectProviderViewModelValidator.Validate(request);
```
Comment (chrisfoster76) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs]: We don't need this explicit call - should be done automatically, and the results added to the model state so we should get them above

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Views/CreateCohort/SelectProvider.cshtml [hunk 8]
```diff
@@ -4,15 +4,42 @@
...
+            </div>
 
+            <details class="govuk-details">
+                <summary class="govuk-details__summary">
+                    <span class="govuk-details__summary-text">
+                        What is a UK Provider Reference Number?
+                    </span>
+                </summary>
+                <div class="govuk-details__text">
+                    A UK Provider Reference Number (UKPRN) is a unique, 8-digit number that identifies a training provider. Your training provider can tell you their UKPRN.
```
Comment (chrisfoster76) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Views/CreateCohort/SelectProvider.cshtml]: We should add the model.ProviderId value into the text box value, so if the user makes a mistake they get to see it in the error state.

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Views/CreateCohort/SelectProvider.cshtml [hunk 9]
```diff
@@ -4,15 +4,42 @@
...
+                </div>
+            </details>
+            <button type="submit" class="govuk-button" tabindex="1">
+                Continue
```
Comment (chrisfoster76) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Views/CreateCohort/SelectProvider.cshtml]: The tab order is out of whack for this element.

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs [hunk 10]
```diff
@@ -35,20 +57,55 @@ public IActionResult Index(IndexRequest request)
...
 
         [Route("select-provider")]
         [HttpPost]
-        public IActionResult SelectProvider(SelectProviderViewModel request)
+        public async Task<IActionResult> SelectProvider(SelectProviderViewModel request)
         {
-            //todo:hit api
+            if (!ModelState.IsValid)
+            {
+                return RedirectToAction("Error", "Error", new { StatusCode = 400 });
```
Comment (chrisfoster76) [inline src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs]: We would want to just re-render the view if the ModelState is not valid, not redirect to an error page

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs (hunk 7 above)
Comment (dashton82) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs]: done

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs (hunk 10 above)
Comment (dashton82) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Controllers/CreateCohortController.cs]: This is done in #29

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Views/CreateCohort/SelectProvider.cshtml (hunk 8 above)
Comment (dashton82) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Views/CreateCohort/SelectProvider.cshtml]: done in #29

File: src/SFA.DAS.EmployerCommitmentsV2.Web/Views/CreateCohort/SelectProvider.cshtml (hunk 9 above)
Comment (dashton82) [reply src/SFA.DAS.EmployerCommitmentsV2.Web/Views/CreateCohort/SelectProvider.cshtml]: done in #29

**Pull request merged**